Bug 1226571 - Add 3D Touch static actions

fluffyemily opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=1226571
* update `Info.plist` to enable static actions
* create `QuickActions` to coordinate/handle quick actions
* add static action handling to `AppDelegate`
* add opening new URL in private mode to `BrowserViewController`
* ensure that when setting `privateMode` in `TabTrayController` in the background that the nil collection view is not called
* add `addTabAndSelect` method for private mode

sleroux commented on Client/Application/AppDelegate.swift:
```diff
@@ -179,6 +201,13 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        if #available(iOS 9, *) {
+            guard let shortcut = quickActions.launchedShortcutItem else { return }
+            quickActions.handleShortCutItem(shortcut, completionBlock: handleShortCutItemType)
+            quickActions.launchedShortcutItem = nil
```
Do we want to add a return statement at the end of this block in case we add more to this method and don't want to passthrough?

fluffyemily replied:
No, but there is an error there. We are returning when there is no launchedShortcutItem and never executing the  `self.profile?.syncManager.applicationDidBecomeActive()` & `self.browserViewController.loadQueuedTabs()`, which means that running iOS9 we will never execute these lines unless we have opened by a shortcut. I'll turn the `guard` into an `if let`. I think we want to execute these lines regardless, for now. If we need to prevent passthrough later we can amend, but right now I can't think of a reason why we would.

sleroux commented on Client/Application/QuickActions.swift:
```diff
@@ -0,0 +1,49 @@
+        completionBlock(type: shortCutType, userData: shortcutItem.userInfo)
+
+        // Construct an alert using the details of the shortcut used to open the application.
+        print("Shortcut Handled: \(shortcutItem.localizedTitle)")
```
Can we use an instance of XCLogger from Loggers.swift instead of print so we can write the log to disk? I think Loggers.browser should work.

fluffyemily replied:
Sorry, that log statement shouldn't be there. Will remove.

sleroux commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -847,13 +847,26 @@ class BrowserViewController: UIViewController {
+        } else {
+            request = nil
+        }
+        let tabTrayController = self.tabTrayController ?? TabTrayController(tabManager: tabManager, profile: profile)
```
Ugh - Can't wait to fix this when I invert the BVC. Storing the private state in the TVC is such a hack :(

sleroux commented:
Not sure if it's because of the simulator plugin I'm using to test 3D touch in the emulator but when triggering the quick action without the app in the background, the app freezes up on launch. When the app is backgrounded everything works fine. I can't attach a debugger to it though since it's a first launch thing :( Are you seeing this behaviour on device?

fluffyemily commented:
I am also seeing that. Looking at the device logs:

`org.mozilla.ios.Fennec failed to scene-create after 18.62s (launch took 1.38s of total time limit 20.00s)

Elapsed total CPU time (seconds): 7.760 (user 7.760, system 0.000), 19% CPU 
Elapsed application CPU time (seconds): 0.141, 0% CPU

Filtered syslog:
None found

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0:
0   libswiftCore.dylib            	0x00000001016044fc 0x1014bc000 + 1344764
1   Client                        	0x00000001002a6ab8 BrowserViewController.browser(Browser, didCreateWebView : WKWebView) -> () (BrowserViewController.swift:1135)
2   Client                        	0x00000001002a77bc @objc BrowserViewController.browser(Browser, didCreateWebView : WKWebView) -> () (BrowserViewController.swift:0)
3   Client                        	0x000000010027491c partial apply + 64
4   Client                        	0x000000010027d184 thunk (Browser.swift:0)
5   Client                        	0x000000010027d1d0 thunk (Browser.swift:0)
6   Client                        	0x000000010027474c Browser.createWebview() -> () (Browser.swift:125)
7   Client                        	0x0000000100260e80 TabManager.configureTab(Browser, request : NSURLRequest?, flushToDisk : Bool, zombie : Bool) -> () (TabManager.swift:258)
8   Client                        	0x00000001002606f0 TabManager.(addTab in _932D52E6383C2B96DB066936B32B3140)(NSURLRequest?, configuration : WKWebViewConfiguration!, flushToDisk : Bool, zombie : Bool, isPrivate : Bool) -> Browser (TabManager.swift:235)
9   Client                        	0x000000010025f9c8 TabManager.addTab(NSURLRequest!, configuration : WKWebViewConfiguration!, isPrivate : Bool) -> Browser (TabManager.swift:182)
10  Client                        	0x000000010025fb0c TabManager.addTabAndSelect(NSURLRequest!, configuration : WKWebViewConfiguration!, isPrivate : Bool) -> Browser (TabManager.swift:187)
11  Client                        	0x000000010029f994 BrowserViewController.openURLInNewTab(NSURL?, isPrivate : Bool) -> () (BrowserViewController.swift:870)
12  Client                        	0x00000001001ce24c AppDelegate.handleShortCutItemType(ShortcutType, userData : [String : NSSecureCoding]?) -> () (AppDelegate.swift:353)
13  Client                        	0x00000001001d07a4 partial apply (AppDelegate.swift:0)
14  Client                        	0x000000010016d260 QuickActions.handleShortCutItem(UIApplicationShortcutItem, completionBlock : (type : ShortcutType, userData : [String : NSSecureCoding]?) -> ()) -> Bool (QuickActions.swift:42)
15  Client                        	0x00000001001cab8c AppDelegate.applicationDidBecomeActive(UIApplication) -> () (AppDelegate.swift:207)
16  Client                        	0x00000001001cae24 @objc AppDelegate.applicationDidBecomeActive(UIApplication) -> () (AppDelegate.swift:0)
17  UIKit                         	0x000000018a0c0a04 -[UIApplication _stopDeactivatingForReason:] + 1080
18  UIKit                         	0x000000018a2e017c -[UIApplication _runWithMainScene:transitionContext:completion:] + 2536
19  UIKit                         	0x000000018a2dcf50 -[UIApplication workspaceDidEndTransaction:] + 168
20  FrontBoardServices            	0x000000018ec0b7c4 -[FBSSerialQueue _performNext] + 184
21  FrontBoardServices            	0x000000018ec0bb44 -[FBSSerialQueue _performNextFromRunLoopSource] + 56
22  CoreFoundation                	0x0000000184a5c544 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
23  CoreFoundation                	0x0000000184a5bfd8 __CFRunLoopDoSources0 + 540
24  CoreFoundation                	0x0000000184a59cd8 __CFRunLoopRun + 724
25  CoreFoundation                	0x0000000184988ca0 CFRunLoopRunSpecific + 384
26  UIKit                         	0x000000018a0a61c8 -[UIApplication _run] + 460
27  UIKit                         	0x000000018a0a0ffc UIApplicationMain + 204
28  Client                        	0x0000000100287b6c main (main.swift:22)
29  libdyld.dylib                 	0x000000019a7928b8 start + 4`

Will look into it....

fluffyemily commented:
The problem is, we are creating the `webView` and attempting to insert it before the `webViewContainer` has been created (before `viewDidLoad` is called on BVC). Therefore it is hanging trying to insert a view at a force unwrapped `UIView` that does not yet exist. This sounds like it's going to be fun to fix :)

fluffyemily commented:
We're left with a problem with applying theme:

STR:
Add PB tab from QuickAction
Return to home screen
Add normal tab from QuickAction

expected: Normal theme has been completely applied
actual: Normal theme has been applied except for TabsButton and BrowserLocationView background which are still PB themed.

Do not yet have a fix for this....

fluffyemily commented:
Fixed a number of issues:

1. Fixed crash when opening tab from cold
          This needed to be done asynchronously otherwise BVC didn't actually exist when tab was created causing crash
2. Fixed failure to apply UIAppearanceTheme
          It seems that once you have applied UIAppearance values to a component, any changes made to those UIAppearance values will not be applied on those components until you remove and then re-add them to a view. Therefore I am forcibly applying UIAppearance themes when starting from a QuickAction.
3. Fixed crash when adding new tab with different mode when tab tray open
        Found a crasher whereby if you opened a tab in a different mode to the one currently in effect in the backgrounded app, while the app is in the TabTray then the collection view works out the incorrect number of cells and crashes when it doesn't match. Fixed by applying new theme just like a tap to the PB button when a collectionView already exists in the TabTray, rather than just updating the privateMode.
4. Feedback from @tecgirl and now new tabs open with the URL bar as first responder.
